File: ui-tests/src/test/java/com/warehouse/ui/pages/LoginPage.java
package com.warehouse.ui.pages;

import com.codeborne.selenide.SelenideElement;
import io.qameta.allure.Step;

import static com.codeborne.selenide.Condition.*;
import static com.codeborne.selenide.Selenide.$;

/**
 * Login Page Object for Warehouse application.
 * Uses data-testid attributes for reliable element selection.
 */
public class LoginPage {

    // Login form elements
    private final SelenideElement usernameInput = $("[data-testid='username-input']");
    private final SelenideElement passwordInput = $("[data-testid='password-input']");
    private final SelenideElement loginButton = $("[data-testid='login-button']");
    private final SelenideElement errorMessage = $("[data-testid='error-message']");

    // After login - navigation elements
    private final SelenideElement logoutButton = $("[data-testid='logout-button']");
    private final SelenideElement navProducts = $("[data-testid='nav-products']");

    @Step("Enter username: {username}")
    public LoginPage enterUsername(String username) {
        usernameInput.shouldBe(visible, enabled).setValue(username);
        return this;
    }

    @Step("Enter password")
    public LoginPage enterPassword(String password) {
        passwordInput.shouldBe(visible, enabled).setValue(password);
        return this;
    }

    @Step("Click login button")
    public LoginPage clickLogin() {
        loginButton.shouldBe(visible, enabled).click();
        return this;
    }

    @Step("Login as {username}")
    public void login(String username, String password) {
        enterUsername(username);
        enterPassword(password);
        clickLogin();
    }

    @Step("Verify error message is displayed")
    public LoginPage verifyErrorDisplayed() {
        errorMessage.shouldBe(visible);
        return this;
    }

    @Step("Verify error message contains: {text}")
    public LoginPage verifyErrorContains(String text) {
        errorMessage.shouldBe(visible).shouldHave(text(text));
        return this;
    }

    @Step("Verify login was successful")
    public void verifyLoginSuccess() {
        logoutButton.shouldBe(visible);
    }

    @Step("Check if logged in")
    public boolean isLoggedIn() {
        return logoutButton.isDisplayed();
    }

    @Step("Verify navigation is visible after login")
    public LoginPage verifyNavigationVisible() {
        navProducts.shouldBe(visible);
        return this;
    }

    @Step("Click logout button")
    public void logout() {
        logoutButton.shouldBe(visible, enabled).click();
    }
}

File: frontend/e2e/pages/login.page.ts
import { Page, Locator } from '@playwright/test'

export class LoginPage {
  readonly page: Page
  readonly usernameInput: Locator
  readonly passwordInput: Locator
  readonly loginButton: Locator
  readonly errorMessage: Locator

  constructor(page: Page) {
    this.page = page
    this.usernameInput = page.locator('[data-testid="username"]')
    this.passwordInput = page.locator('[data-testid="password"]')
    this.loginButton = page.locator('[data-testid="login-button"]')
    this.errorMessage = page.locator('[data-testid="error-message"]')
  }

  async goto() {
    await this.page.goto('/login')
    await this.usernameInput.waitFor({ state: 'visible', timeout: 1000 })
  }

  async login(username: string, password: string) {
    await this.usernameInput.fill(username)
    await this.passwordInput.fill(password)
    await this.loginButton.click()
  }

  async getErrorMessage(): Promise<string> {
    return await this.errorMessage.textContent() || ''
  }

  async isErrorVisible(): Promise<boolean> {
    try {
      await this.errorMessage.waitFor({ state: 'visible', timeout: 1000 })
      return true
    } catch {
      return false
    }
  }
}
